### PR TITLE
Centralizedconfiguration uses the parameter list as a group list when executing commands.

### PR DESCRIPTION
**Pull request by Nicogp**
|Related issue|
|---|
|#234|

## Description

This PR allows the CentralizedConfiguration component to receive the parameter list(group list) for command execution.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] MAC OS X

### CODE REVIEW

**Review by jr0me — COMMENTED** [review]
I'm confused about the conversion between json and strings that we are making with the parameters. If we are allowing some of the strings in the parameters to be json objects, then it makes little sense to use a vector of strings if we have to parse and convert them back to json, and parse them again...

Edit: we also need to modify the `ModuleWrapper`, and the `ExecuteCommand` from modules that implement that wrapper, that is missing in the pr.

File: src/agent/command_store/src/command_store.cpp [hunk 1]
```diff
@@ -150,7 +155,8 @@ namespace command_store
                 }
                 else if (col.Name == "parameters")
                 {
-                    cmd.Parameters = col.Value;
+                    auto deserializedJsonArray = nlohmann::json::parse(col.Value);
```
[the remark this answers is not in the document]
Comment (Nicogp) [reply src/agent/command_store/src/command_store.cpp]: Done!

File: src/agent/command_store/src/command_store.cpp [hunk 2]
```diff
@@ -182,7 +188,10 @@ namespace command_store
         if (!cmd.Command.empty())
             fields.emplace_back("command", sqlite_manager::ColumnType::TEXT, cmd.Command);
         if (!cmd.Parameters.empty())
-            fields.emplace_back("parameters", sqlite_manager::ColumnType::TEXT, cmd.Parameters);
+        {
+            nlohmann::json serializedJsonArray = cmd.Parameters;
```
[the remark this answers is not in the document]
Comment (Nicogp) [reply src/agent/command_store/src/command_store.cpp]: Done!

File: src/agent/command_store/tests/command_store_test.cpp [hunk 3]
```diff
@@ -25,35 +25,70 @@ TEST_F(CommandStoreTest, StoreCommandTest)
...
+    m_commandStore->Clear();
+
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+    module_command::CommandEntry cmd1(TESTID_5,
+                                      "Module1",
+                                      "{CommandTextHERE}",
+                                      {"Parameter1", "Parameter 2", "3"},
+                                      "Result1",
+                                      module_command::Status::IN_PROGRESS);
+    bool retVal = m_commandStore->StoreCommand(cmd1);
```
[the remark this answers is not in the document]
Comment (Nicogp) [reply src/agent/command_store/tests/command_store_test.cpp]: Done!

File: src/agent/command_store/tests/command_store_test.cpp [hunk 4]
```diff
@@ -25,35 +25,70 @@ TEST_F(CommandStoreTest, StoreCommandTest)
...
+    ASSERT_EQ(retVal, true);
```
[the remark this answers is not in the document]
Comment (Nicogp) [reply src/agent/command_store/tests/command_store_test.cpp]: Done!

File: src/agent/centralized_configuration/include/centralized_configuration.hpp [hunk 5]
```diff
@@ -22,9 +22,11 @@ namespace centralized_configuration
         /**
          * @brief Executes a command for the centralized configuration system.
          * @param command A string containing a JSON command to execute.
+         * @param parameters A vector of strings containing the parameters of the command to be execute.
```
[the remark this answers is not in the document]
Comment (Nicogp) [reply src/agent/centralized_configuration/include/centralized_configuration.hpp]: Done!

File: src/agent/centralized_configuration/src/centralized_configuration.cpp [hunk 6]
```diff
@@ -8,18 +8,23 @@
...
             {
                 if (m_setGroupIdFunction && m_downloadGroupFilesFunction)
                 {
-                    const auto groupIds = commnandAsJson["groups"].get<std::vector<std::string>>();
+                    if (parameters.empty())
+                        co_return module_command::CommandExecutionResult {
+                            module_command::Status::FAILURE,
+                            "CentralizedConfiguration group set failed, no group list"};
+
+                    const auto parametersAsJson = nlohmann::json::parse(parameters[0]);
```
[the remark this answers is not in the document]
Comment (jr0me) [reply src/agent/centralized_configuration/src/centralized_configuration.cpp:26]: Shouldn't we just use a vector of json objects? We are converting them to strings, only to convert them back to json objects, then we convert them back...

File: src/agent/centralized_configuration/include/centralized_configuration.hpp [hunk 7]
```diff
@@ -22,9 +24,11 @@ namespace centralized_configuration
         /**
          * @brief Executes a command for the centralized configuration system.
          * @param command A string containing a JSON command to execute.
+         * @param parameters A vector of json containing the parameters of the command to be executed.
```
Comment (TomasTurina) [inline src/agent/centralized_configuration/include/centralized_configuration.hpp:27]: Fix this description.

File: src/agent/centralized_configuration/src/centralized_configuration.cpp [hunk 8]
```diff
@@ -6,18 +6,23 @@
 
 namespace centralized_configuration
 {
-    boost::asio::awaitable<module_command::CommandExecutionResult>
-    CentralizedConfiguration::ExecuteCommand(const std::string command) // NOLINT(performance-unnecessary-value-param)
+    boost::asio::awaitable<module_command::CommandExecutionResult> CentralizedConfiguration::ExecuteCommand(
+        const std::string command,       // NOLINT(performance-unnecessary-value-param)
```
Comment (TomasTurina) [inline src/agent/centralized_configuration/src/centralized_configuration.cpp:10]: Why are these `NOLINT` for?

File: src/agent/centralized_configuration/src/centralized_configuration.cpp [hunk 9]
```diff
@@ -37,7 +42,7 @@ namespace centralized_configuration
                         module_command::Status::FAILURE, "CentralizedConfiguration group set failed, no function set"};
                 }
             }
-            else if (commnandAsJson["command"] == "update-group")
+            else if (command == "update-group")
```
Comment (TomasTurina) [inline src/agent/centralized_configuration/src/centralized_configuration.cpp:45]: We should also log something if the command is not recognized.

File: src/agent/centralized_configuration/src/centralized_configuration.cpp [hunk 10]
```diff
@@ -6,18 +6,23 @@
...
                 {
-                    const auto groupIds = commnandAsJson["groups"].get<std::vector<std::string>>();
+                    if (parameters.empty())
+                        co_return module_command::CommandExecutionResult {
+                            module_command::Status::FAILURE,
+                            "CentralizedConfiguration group set failed, no group list"};
+
+                    const auto groupIds = parameters[0].get<std::vector<std::string>>();
 
                     m_setGroupIdFunction(groupIds);
```
Comment (TomasTurina) [inline src/agent/centralized_configuration/src/centralized_configuration.cpp:28]: Code after this line is duplicated, maybe we can move it after the `if/else` statement.

File: src/agent/centralized_configuration/src/centralized_configuration.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nicogp) [reply src/agent/centralized_configuration/src/centralized_configuration.cpp]: Done!

File: src/agent/centralized_configuration/include/centralized_configuration.hpp (hunk 7 above)
Comment (Nicogp) [reply src/agent/centralized_configuration/include/centralized_configuration.hpp]: Done!

File: src/agent/centralized_configuration/src/centralized_configuration.cpp (hunk 8 above)
Comment (Nicogp) [reply src/agent/centralized_configuration/src/centralized_configuration.cpp:10]: Due to a performance warning related to parameter passing by value. Previously it was for the command parameter and now it was added also for “parameters”.

File: src/agent/centralized_configuration/src/centralized_configuration.cpp (hunk 9 above)
Comment (Nicogp) [reply src/agent/centralized_configuration/src/centralized_configuration.cpp:45]: There are no log messages anywhere in the component, we could create an issue to review where they need to be added or do it as part of the following tasks. What do you think?

File: src/agent/centralized_configuration/include/centralized_configuration.hpp [hunk 11]
```diff
@@ -4,6 +4,8 @@
 
 #include <boost/asio/awaitable.hpp>
 
+#include <nlohmann/json.hpp>
```
Comment (jr0me) [inline src/agent/centralized_configuration/include/centralized_configuration.hpp:7]: Now it should be added as a public dependency and linked against it in the cmake file of centralized configuration.

File: src/agent/module_command/include/module_command/command_entry.hpp [hunk 12]
```diff
@@ -1,5 +1,7 @@
 #pragma once
 
+#include <nlohmann/json.hpp>
```
Comment (jr0me) [inline src/agent/module_command/include/module_command/command_entry.hpp:3]: Same comment as for centralized configuration.

> Now it should be added as a public dependency and linked against it in the cmake file of centralized configuration.

**Review by jr0me — COMMENTED** [review]
Looking good.

The updates to the `ExecuteCommand` functions are still missing.

File: src/agent/src/command_handler_utils.cpp [hunk 13]
```diff
@@ -7,6 +7,8 @@
 #include <boost/system/system_error.hpp>
 #include <logger.hpp>
 
+#include <nlohmann/json.hpp>
```
[the remark this answers is not in the document]
Comment (Nicogp) [reply src/agent/src/command_handler_utils.cpp]: Done!

File: src/agent/centralized_configuration/include/centralized_configuration.hpp (hunk 11 above)
Comment (Nicogp) [reply src/agent/centralized_configuration/include/centralized_configuration.hpp:7]: Done!

File: src/agent/module_command/include/module_command/command_entry.hpp (hunk 12 above)
Comment (Nicogp) [reply src/agent/module_command/include/module_command/command_entry.hpp:3]: Done!

**Comment by Nicogp** [comment]
@jr0me 

> The updates to the ExecuteCommand functions are still missing.

Done!

**Review by jr0me — APPROVED** [review]
LGTM

File: src/modules/include/moduleManager.hpp [hunk 14]
```diff
@@ -33,9 +33,9 @@ class ModuleManager {
             .Start = [&module]() { module.Start(); },
             .Setup = [&module](const configuration::ConfigurationParser& configurationParser) { module.Setup(configurationParser); },
             .Stop = [&module]() { module.Stop(); },
-            .ExecuteCommand = [&module](std::string query) -> Co_CommandExecutionResult
+            .ExecuteCommand = [&module](std::string query, nlohmann::json parameters) -> Co_CommandExecutionResult
```
Comment (TomasTurina) [inline src/modules/include/moduleManager.hpp:36]: I think it will be clearer to name the first parameter `command`. Change this in all modules.

Comment (jr0me) [reply src/modules/include/moduleManager.hpp:36]: I also noted that you are using `[[maybe_unused]]` in function declarations. They only make sense in function definitions. If possible I would take them out, though it doesn't make a big difference.

Comment (Nicogp) [reply src/modules/include/moduleManager.hpp]: Done!

**Pull request merged**